src/deserialize.rs

```diff
@@ -0,0 +1,405 @@
+    /// and/or [`DeserializeFlags::RESIZEABLE`].
+    ///
+    /// See the C Interface Specification [Deserialize a database](https://www.sqlite.org/c3ref/deserialize.html).
+    pub unsafe fn deserialize_with_flags(
+        &mut self,
+        db: DatabaseName<'_>,
+        data: &[u8],
+        cap: usize,
+        flags: DeserializeFlags,
+    ) -> Result<()> {
+        let path = &mut self.path;
+        let mut c = self.db.borrow_mut();
+        let db = db.to_cstring()?;
+        let rc = ffi::sqlite3_deserialize(
+            c.db(),
+            db.as_ptr(),
+            data.as_ptr() as *mut _,
+            data.len() as _,
+            cap as _,
+            flags.bits() as _,
+        );
+        c.decode_result(rc).map(|_| {
+            *path = Some(":memory:".into());
+        })
+    }
+
+    /// Returns the serialization of a database, or `None` when [`DatabaseName`] does not exist.
+    pub fn serialize(&self, db: DatabaseName<'_>) -> Result<Option<SerializedDb>> {
+        unsafe { self.serialize_with_flags(db, SerializeFlags::empty()) }
+    }
+
+    /// The serialization is the same sequence of bytes which would be written to disk
+    /// if the database where backed up to disk.
+    ///
+    /// Returns `Ok(None)` when [`DatabaseName`] does not exist or malloc/prepare fails.
+    ///
+    /// # Safety
+    ///
+    /// If [`SerializeFlags::NO_COPY`] is set, this returns a pointer to memory that SQLite is currently using
+    /// (or `Ok(None)` if this is not available).
+    /// In that case, the returned `SerializedDb` mutably borrows `Connection`,
+    /// [`std::mem::forget()`] one of them to prevent double free.
+    ///
+    /// See the C Interface Specification [Serialize a database](https://www.sqlite.org/c3ref/serialize.html).
+    pub unsafe fn serialize_with_flags(
+        &self,
+        db: DatabaseName<'_>,
+        flags: SerializeFlags,
+    ) -> Result<Option<SerializedDb>> {
+        let c = self.db.borrow();
+        let db = db.to_cstring()?;
+        let mut size = 0;
+        let data = ffi::sqlite3_serialize(
+            c.db(),
+            db.as_ptr(),
+            &mut size as *mut _ as *mut _,
+            flags.bits() as _,
+        );
+        Ok(NonNull::new(data).map(|data| SerializedDb {
+            data,
+            len: size,
+            cap: size,
+            phantom: PhantomData,
+        }))
+    }
+}
+
+/// Serialized database content (a growable vector of `u8`) owned by SQLite.
+/// Used for [`Connection::serialize`] and [`Connection::deserialize`].
+/// Memory allocation is handled by `sqlite3_malloc64`, `sqlite3_realloc64`, `sqlite3_msize` and `sqlite3_free`.
+/// ```
+/// # use rusqlite::deserialize::SerializedDb;
+/// let mut serialized = SerializedDb::new();
+/// serialized.extend(vec![1, 2, 3]);
+/// assert_eq!(serialized[1], 2);
+/// ```
+pub struct SerializedDb {
+    data: NonNull<u8>,
+    len: usize,
+    cap: usize,
+    phantom: PhantomData<[u8]>,
```

Comment:
What's the point of `PhantomData<[u8]>`?

Reply:
You're right, `u8` does not implement `Drop` so the drop checker should not worry about it.